dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java

```diff
@@ -1095,9 +1096,45 @@ public void endVisit(Initializer x, MethodScope scope) {
     public void endVisit(InstanceOfExpression x, BlockScope scope) {
       try {
         SourceInfo info = makeSourceInfo(x);
-        JExpression expr = pop(x.expression);
+        JExpression expr;
+        JDeclarationStatement jDeclarationStatement = null;
+        if (x.pattern != null) {
+          jDeclarationStatement = (JDeclarationStatement) pop();
+        }
+        expr = pop(x.expression);
         JReferenceType testType = (JReferenceType) typeMap.get(x.type.resolvedType);
-        push(new JInstanceOf(info, testType, expr));
+
+        if (jDeclarationStatement == null) {
+          push(new JInstanceOf(info, testType, expr));
+        } else {
+          String localVariableName = jDeclarationStatement.getVariableRef().getTarget().getName();
+          if (!curMethod.instanceOfDeclarations.containsKey(localVariableName)) {
+            curMethod.body.getBlock().addStmt(0, jDeclarationStatement);
+            curMethod.instanceOfDeclarations.put(localVariableName, jDeclarationStatement);
+          }
+
+          // rewrite (o instanceof Foo foo)
+          // to
+          // Foo foo;
+          // (o instanceof Foo && null != (foo = (Foo) o))
+          CloneExpressionVisitor cloner = new CloneExpressionVisitor();
+
+          JVariableRef variableRef = jDeclarationStatement.getVariableRef();
+          JCastOperation jCastOperation = new JCastOperation(info, variableRef.getType(), cloner.cloneExpression(expr));
```

Comment:
Remember that cloning this will produce unwanted side effects - see my last comment on the PR. Instead we want another local variable.

If `!expr.hasSideEffects()`, then what you have is shorter and better I think. (not sure if the compiler will optimize that automatically).